### PR TITLE
Refactor Object and Class

### PR DESCRIPTION
**Pull request by st0012**
Simplify object and class' design and try to follow Golang's convention like using getter instead of expose type.

### CODE REVIEW

File: vm/array.go [hunk 1]
```diff
@@ -43,9 +43,9 @@ func (a *ArrayObject) toJSON() string {
 	return out.String()
 }
 
-// returnClass returns current object's class, which is RArray
-func (a *ArrayObject) returnClass() Class {
-	return a.Class
+// Class returns current object's class, which is RArray
+func (a *ArrayObject) Class() *RClass {
```
Comment (st0012) [inline vm/array.go]: This is the `getter` method I am talking about.

File: vm/object.go [hunk 2]
```diff
@@ -22,26 +22,30 @@ func (p *Pointer) returnClass() *RClass {
 
 // RObject represents any non built-in class's instance.
 type RObject struct {
-	Class             *RClass
+	*baseObj
+	InitializeMethod *MethodObject
+}
+
+type baseObj struct {
```
Comment (st0012) [inline vm/object.go]: By embedding `baseObj`, every object can have consist attributes like class and instance variables.

File: vm/instruction.go [hunk 3]
```diff
@@ -256,14 +256,14 @@ var builtInActions = map[operationType]*action{
...
 
-			method := &MethodObject{Name: methodName, argc: argCount, instructionSet: is, class: t.vm.builtInClasses["Method"]}
+			method := &MethodObject{Name: methodName, argc: argCount, instructionSet: is, baseObj: &baseObj{class: t.vm.builtInClasses[methodClass]}}
 
 			v := t.stack.pop().Target
 			switch self := v.(type) {
 			case *RClass:
 				self.Methods.set(methodName, method)
 			default:
-				self.returnClass().(*RClass).Methods.set(methodName, method)
```
Comment (st0012) [inline vm/instruction.go]: This can also reduce some type assertions.

**Comment by hachi8833** [comment]
Good refactoring!
Actually, yesterday I suppose that `RClass` might be redundant, but I was not confident about this...

File: vm/object.go (hunk 2 above)
Comment (st0012) [inline vm/object.go]: And from now on,  if objects' function use same attribute we just need to implement it on `baseObj` once. Just like `instanceVariableGet` and `instanceVariableSet`

**Comment by st0012** [comment]
@hachi8833 You can just point out what you think should be improved 😄

**Pull request merged**